src/librustc/ty/wf.rs

```diff
@@ -346,6 +346,21 @@ impl<'a, 'gcx, 'tcx> WfPredicates<'a, 'gcx, 'tcx> {
                     // WFedness.)
                 }
 
+                ty::TyClosure(def_id, substs) => {
+                    // Just check the upvar types for WF. This is
+                    // needed because we capture the signature and it
+                    // may not be WF without the implied
+                    // bounds. Consider a closure like `|x: &'a T|` --
+                    // it may be that `T: 'a` is not known to hold in
+                    // the creator's context (and indeed the closure
+                    // may not be invoked by its creator, but rather
+                    // turned to someone who *can* verify that).
+                    subtys.skip_current_subtree(); // subtree handled by compute_projection
+                    for upvar_ty in substs.upvar_tys(def_id, self.infcx.tcx) {
+                        self.compute(upvar_ty);
+                    }
+                }
+
```

Comment:
Shouldn't this code also apply to generators?

Reply:
I should improve the comment. Generators are currently undergoing a *stricter* check, in which all reachable types are checked, not only the upvar types. This is sort of irrelevant to generators anyway since they don't capture their signature, just the yield type -- and since there are no implied bounds to consider, the WF of the yield type should follow directly from the WF of the other parts of the generator. 

The special treatment of closures here really ought not to be necessary either; the problem is related to https://github.com/rust-lang/rust/issues/25860 -- there is no way for us to express a `fn` type complete with the implied bounds that it is assuming. I think in reality the WF rules around `fn` are a bit messed up, and that is the rot problem: `fn(&'a T)` should probably always be WF, because it should be shorthand for something like `where(T: 'a) { fn(&'a T) }`, as discussed in #25860.